File: HW8/SourceFiles/GTUSet.java
/**
 * The type GTUSet
 *
 * @param <T> the type parameter
 */
public class GTUSet<T> extends GTUContainer<T>{
    /**
     * Instantiates a new GTUSet(Empty).
     */
    public GTUSet() {
        data = null;
        size = 0;
    }

    /**
     * Instantiates a new GTUSet(with One data).
     *
     * @param _data the data
     */
    public GTUSet(T _data) {
        size = 1;
        T[] temp = (T[]) new Object[size];
        temp[0] = _data;
        data = temp;
    }

    /**
     * Instantiates a new GTUSet(with T array).
     *
     * @param _data the data
     */
    public GTUSet(T[] _data) {
        data = _data;
        size = _data.length;
    }

    @Override
    public void insert(T element) {
        try {
            if(size >= max_size)
                throw new Exception("Maximum size reached. No more elements can be added.");
            boolean cntrl = true;
            T[] temp;
            for (int i = 0; i < size; i++)
                if (element.equals(data[i]))
                    cntrl = false;

            if(!cntrl)
                throw new Exception("The element failed to add since it is already in the set.");

            if (cntrl) {
                temp = (T[]) new Object[size + 1];
            } else {
                temp = (T[]) new Object[size];
            }
            for (int i = 0; i < size; i++) {
                temp[i] = data[i];
            }
            if (cntrl)
                temp[size++] = element;

            data = temp;
        }
        catch(Exception e){
            System.out.println(e.getMessage());
        }
    }

    @Override
    public void insertArray(T[] array) {
        boolean cntrl=true;
        T[] temp;
        for(int j=0;j<array.length;++j){
            try {
                if(size >= max_size)
                    throw new Exception("Maximum size reached. No more elements can be added.");
                for (int i = 0; i < size; i++)
                    if (array[j].equals(data[i]))
                        cntrl = false;

                if(!cntrl)
                    throw new Exception("The element failed to add since it is already in the set.");

                if (cntrl) {
                    temp = (T[]) new Object[size + 1];
                } else {
                    temp = (T[]) new Object[size];
                }
                for (int i = 0; i < size; i++) {
                    temp[i] = data[i];
                }
                if (cntrl)
                    temp[size++] = array[j];

                data = temp;
            }
            catch(Exception e){
                System.out.println(e.getMessage());
            }
        }
    }

    @Override
    public void erase(T element) {
        T[] temp = (T[]) new Object[size-1];

        for(int i=0,k=0;i<size;i++,k++){
            if(data[i] != element){
                temp[k] = data[i];
            }else{
                k--;
            }
        }
        size--;
        data = temp;
    }

    @Override
    public GTUIterator<T> iterator() {
        return new GTUIterator<T>(data);
    }

}

File: HW6/GTUSet.cpp
#include<iostream>
#include"GTUSet.h"
#include"GTUIterator.h"
#include"GTUIteratorConst.h"

#define SIZE 9999

namespace GTUContainers{
    using namespace std;

    template<typename T>
    GTUSet<T>::GTUSet() : GTUContainer<T>(){
        data = nullptr;
        count = 0;
    }

    template<typename T>
    GTUSet<T>::GTUSet(T ptr) : GTUContainer<T>(){
        (*data)[0] = ptr;
        count = 1;
    }

    template<typename T>
    bool GTUSet<T>::empty(){
        return (count == 0);
    }

    template<typename T>
    int GTUSet<T>::size(){
        return count;
    }

    template<typename T>
    int GTUSet<T>::max_size(){
        return SIZE; 
    }

    template<typename T>
    void GTUSet<T>::insert(const T& new_element){
        if(count < SIZE){
            if(data == nullptr){
                data = make_shared<T*>(new T[SIZE]);
                (*data)[0] = new_element;
                count = 1;
                return;
            }
            T temp;
            int cntrl=1;
            for(int i=0;i<count;i++){
                if(new_element == (*data)[i]){
                    cntrl=0;
                }else if(i == count-1 && cntrl){
                    cntrl=1;
                }
            }    
            if(cntrl){
                (*data)[count] = new_element;
                count++; 
                for (int i=0;i<count-1;i++){
                    for (int j=0;j<count-i-1;j++){
                        if((*data)[j] > (*data)[j+1]){
                            temp =  (*data)[j];
                            (*data)[j] = (*data)[j+1];
                            (*data)[j+1] = temp;
                        }
                    }
                }
            }else{
                try{
                    throw "\nERROR : The same number already exists.\n";
                }
                catch(const char *error){
                    cout << error;
                }
            }
        }   
    }

    template<typename T>
    void GTUSet<T>::erase(const T& element){
        int cntrl=0,cntrl2=1;
        for(int i=0,k=0;i<count;i++,k++){
            if(element == (*data)[i]){
                if(cntrl2)
                    k++;
                cntrl=1;
                cntrl2=0;
            }
            (*data)[i] = (*data)[k];
        }
        if(cntrl){
            (*data)[count-1] = 0;
            count--;
        }
    }

    template<typename T>
    void GTUSet<T>::clear(){
        for(int i=0;i<count;i++){
            (*data)[i] = 0;
        }
        data = nullptr;
        count = 0;
    }

}